### PR TITLE
Stop guessing entity names for patterns of multi-pattern resources

### PR DESCRIPTION
**Pull request by hzyi-google**
Stop creating resource name components on the fly.
- Stop generating parent resources on the fly. Instead, explicitly look for a parent resource in all the resources defined.
- Stop generating entity names for each single pattern of a multi-pattern resource name. Instead, use entity names defined in `deprecated_collections` in gapic v2.
- Converting common resources to file-level resource definitions.

### CODE REVIEW

File: src/test/java/com/google/api/codegen/protoannotations/testdata/java_library.baseline [hunk 1]
```diff
@@ -4500,19 +4497,6 @@ public class LibraryClient implements BackgroundResource {
...
       PathTemplate.createWithoutUrlEncoding("shelves/{shelf_id}");
 
-  /**
-   * Formats a string containing the fully-qualified path to represent
-   * a archive_book resource.
-   *
-   * @deprecated Use the {@link ArchiveBookName} class instead.
-   */
-  @Deprecated
-  public static final String formatArchiveBookName(String archive, String book) {
```
Comment (hzyi-google) [inline src/test/java/com/google/api/codegen/protoannotations/testdata/java_library.baseline]: The `archive_book` resource name was created through the "Trie" strategy. We stop generate any resource name helpers for this pattern.

File: src/test/java/com/google/api/codegen/protoannotations/testdata/java_library_no_gapic_config.baseline [hunk 2]
```diff
@@ -1070,7 +1070,7 @@ public class LibraryServiceClient implements BackgroundResource {
    * Sample code:
    * <pre><code>
    * try (LibraryServiceClient libraryServiceClient = LibraryServiceClient.create()) {
-   *   BookName name = BookName.of("[BOOK_SHELF]", "[BOOK]");
```
Comment (hzyi-google) [inline src/test/java/com/google/api/codegen/protoannotations/testdata/java_library_no_gapic_config.baseline]: `Book` is a multi-pattern resource. Without gapic v2, the generator does not know how to instantiate an instance of it, so we use a string literal directly.

**Comment by hzyi-google** [comment]
PTAL

**Comment by hzyi-google** [comment]
Diffs as an example:

generated bigquery datatransfer clients using gapic v1 + gapic generator master branch and gapic v2 + this PR: https://github.com/hzyi-google/multi-pattern-resource-names/compare/bqdt-only-deprecated?expand=1

File: src/test/java/com/google/api/codegen/protoannotations/testdata/java_library.baseline [hunk 3]
```diff
@@ -14679,15 +14641,15 @@ public class LibraryClientTest {
...
+    ResourceName stringsElement = ArchivedBookName.of("[ARCHIVE]", "[BOOK]");
     List<ResourceName> strings = Arrays.asList(stringsElement);
     ListStringsResponse expectedResponse = ListStringsResponse.newBuilder()
       .setNextPageToken(nextPageToken)
       .addAllStrings(UntypedResourceName.toStringList(strings))
       .build();
     mockLibraryService.addResponse(expectedResponse);
 
-    ResourceName name = ArchiveBookName.of("[ARCHIVE]", "[BOOK]");
+    ResourceName name = ArchivedBookName.of("[ARCHIVE]", "[BOOK]");
```
Comment (vam-google) [inline src/test/java/com/google/api/codegen/protoannotations/testdata/java_library.baseline]: Putting this comment here to note our offline the discussions result:
`ArchiveBookName` was changed to `ArchivedBookName` as a result of fixing basically broken configuration used for the tests

**Comment by hzyi-google** [comment]
I confirm both statements are correct.

File: src/main/java/com/google/api/codegen/config/ResourceDescriptorConfig.java [hunk 4]
```diff
@@ -106,130 +109,120 @@ public static ResourceDescriptorConfig from(
...
+              getUnqualifiedTypeName(),
+              getPatterns().get(0),
+              configOverrides.get(
+                  unqualifiedTypeName.toLowerUnderscore()))); // entity names in gapic config are in
+      // lower_underscore case.
+    }
+    ImmutableMap.Builder<String, ResourceNameConfig> resourceNameConfigs = ImmutableMap.builder();
+
+    // Multi-pattern resource.
+    for (String pattern : getPatterns()) {
```
[the remark this answers is not in the document]
Comment (hzyi-google) [reply src/main/java/com/google/api/codegen/config/ResourceDescriptorConfig.java]: Line 128 checks that `getPatterns()` cannot be zero.

File: src/main/java/com/google/api/codegen/config/GapicProductConfig.java [hunk 5]
```diff
@@ -787,25 +821,59 @@ private static GapicProductConfig createDummyInstance(
...
+      // - parent resources of those referenced by any message in source protos through `child_type`
+
+      // Message-level resource definitions.
+      if (resourceDescriptorConfig.isDefinedAtMessageLevel()
+          && resourceDescriptorConfig
+              .getAssignedProtoFile()
+              .getProto()
+              .getPackage()
+              .equals(defaultPackage)) {
+        Map<String, ResourceNameConfig> resources =
```
[the remark this answers is not in the document]
Comment (hzyi-google) [reply src/main/java/com/google/api/codegen/config/GapicProductConfig.java]: Good point!

Although there is a chance that a resource is both referenced by `type` and `child_type`, in which case we need to call `buildResourceNameConfigs` on both `annotationResourceNameConfig` and `actualDescConfig`, so we can't squash the latter two `if`s.

I did combined the former two `if`s into one.

File: src/main/java/com/google/api/codegen/config/ResourceNameMessageConfigs.java [hunk 6]
```diff
@@ -121,49 +84,73 @@ static ResourceNameMessageConfigs createFromAnnotations(
...
-        entityName = config.getDerivedParentEntityName();
+    String entityName =
+        ResourceDescriptorConfig.getUnqualifiedTypeName(resourceDescriptor.getType());
+    if (resourceDescriptor.getPatternList().size() > 1) {
+      entityName = entityName + "Oneof";
+    }
+    fieldEntityMap.put(resourceFieldName, entityName);
+  }
+
+  private static void collectFieldEntityMapForResourceReference(
```
[the remark this answers is not in the document]
Comment (hzyi-google) [reply src/main/java/com/google/api/codegen/config/ResourceNameMessageConfigs.java]: Updated to `loadFieldEntityPairFromResourceAnnotation` and `loadFieldEntityPairFromResourceReferenceAnnotation` and added descriptions.

File: src/main/java/com/google/api/codegen/transformer/java/JavaFeatureConfig.java [hunk 7]
```diff
@@ -35,10 +37,33 @@ public boolean resourceNameTypesEnabled() {
...
+    // deprecated_collections
+    // in gapic config v2 so that the generator can create resource name strings in the old way, but
+    // not
+    // for this specific test without a gapic config.
+    boolean requiresMultiPatternResourceSupport =
+        fieldConfig.getResourceNameType() == ResourceNameType.ONEOF
+            && ((ResourceNameOneofConfig) fieldConfig.getResourceNameConfig())
+                    .getSingleResourceNameConfigs()
+                    .size()
+                == 0;
```
[the remark this answers is not in the document]
Comment (hzyi-google) [reply src/main/java/com/google/api/codegen/transformer/java/JavaFeatureConfig.java]: change to `isEmpty()` (Don't know why I did the size check instead :P)

File: src/main/java/com/google/api/codegen/config/ResourceDescriptorConfig.java [hunk 8]
```diff
@@ -106,130 +109,120 @@ public static ResourceDescriptorConfig from(
...
+
+      parentResourceCandidates.retainAll(
+          patternResourceDescriptorMap.getOrDefault(parentPattern, Collections.emptySet()));
     }
+
+    // Filter out resources that have more patterns than childResource does.
+    parentResourceCandidates =
+        parentResourceCandidates
+            .stream()
+            .filter(c -> c.getPatterns().size() == childResource.getPatterns().size())
```
[the remark this answers is not in the document]
Comment (hzyi-google) [reply src/main/java/com/google/api/codegen/config/ResourceDescriptorConfig.java]: Added.

**Comment by hzyi-google** [comment]
PTAL

**Pull request merged**